core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -164,6 +172,14 @@ object RequestChannel extends Logging {
         m.totalTimeHist.update(totalTime)
       }
 
+      // Record network handler thread usage. This is included towards the request quota for
+      // the user/client. Throttling is only performed when request handler thread usage
+      // is recorded, just before responses are queued for delivery.
+      recordNetworkThreadUsageCallback match {
+        case Some(record) => record(math.max(endTimeNanos - responseDequeueTimeNanos, 0))
```

Comment:
Hmm, not sure if this is accurate to capture the network thread utilization. What we are recording is essentially the responseSendTime, which includes the time for waiting for the socket to be writable. That portion of the time actually doesn't tie up the network threads and shouldn't be accounted for in the request time. Also, this seems to only cover the network thread time for sending responses, not for reading requests (which could be significant for produce requests).

I was thinking that we may need to do the following. In Selector.pollSelectionKeys(), we will measure the time spent for reading/writing each KafkaChannel and propagate the time back to the caller. Then, we can account for both request/response time in network threads in SocketServer.